### ISSUE TITLE
ConnectAsync_CancellationRequestedAfterConnect_ThrowsOperationCanceledException

### ISSUE THREAD

**Issue by ManickaP** [dotnet/runtime#41511]
### System.Net.WebSockets.Client.Tests

**System.Net.NetworkException : The operation is not allowed on non-connected sockets.**

net5.0-OSX-Debug-x64-Mono_release-OSX.1014.Amd64.Open

```
System.Net.NetworkException : The operation is not allowed on non-connected sockets.
   at System.Net.Sockets.NetworkStream..ctor(Socket socket, FileAccess access, Boolean ownsSocket) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs:line 81
   at System.Net.Sockets.NetworkStream..ctor(Socket socket, Boolean ownsSocket) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs:line 32
   at System.Net.Test.Common.LoopbackServer.EstablishConnectionAsync() in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 115
   at System.Net.Test.Common.LoopbackServer.AcceptConnectionAsync(Func`2 funcAsync) in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 121
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 96
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 124
   at System.Net.Test.Common.LoopbackServer.<>c__DisplayClass13_0.<<CreateClientAndServerAsync>b__0>d.MoveNext() in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 87
--- End of stack trace from previous location ---
   at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options) in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 71
   at System.Net.WebSockets.Client.Tests.ConnectTest.ConnectAsync_CancellationRequestedAfterConnect_ThrowsOperationCanceledException() in /_/src/libraries/System.Net.WebSockets.Client/tests/ConnectTest.cs:line 264
--- End of stack trace from previous location ---
```

https://dev.azure.com/dnceng/public/_build/results?buildId=790029&view=ms.vss-test-web.build-test-results-tab&runId=24931714&resultId=130442&paneView=debug

Hit in https://github.com/dotnet/runtime/pull/41443

**Comment by geoffkizer** [comment]
@ManickaP I thought this got fixed? Is this a separate issue?

**Comment by jaredpar** [comment]
One failure in non-PR builds in last 14 days. We should consider closing this one. 

https://runfo.azurewebsites.net/search/tests/?bq=definition%3Aruntime+started%3A%7E14+kind%3A%21pr&tq=name%3AConnectAsync_CancellationRequestedAfterConnect_ThrowsOperationCanceledException

**Comment by jaredpar** [comment]
That is because there are no failures in last 14 days now. Basically the one failure was right on the edge and by the time you clicked it was gone. The site didn't do a good job of displaying "no results".

**Comment by lewing** [comment]
Failed today on `release/5.0`

**Comment by jaredpar** [comment]
@lewing  can u provide a link to the build with the failure? Looking at the latest build against the linked PR I do not see this failure 

https://dev.azure.com/dnceng/public/_build/results?buildId=846382&view=ms.vss-test-web.build-test-results-tab

**Comment by lewing** [comment]
@jaredpar https://dev.azure.com/dnceng/public/_build/results?buildId=845440&view=logs&jobId=36e79fca-96b2-574d-d647-cc0884a2a88a&j=36e79fca-96b2-574d-d647-cc0884a2a88a&t=e30f6ac0-d99b-5629-c150-05bcfed26194

**Comment by danmosemsft** [comment]
This failed on a release/5.0 PR
```
System.Net.WebSockets.Client.Tests.ConnectTest.ConnectAsync_CancellationRequestedAfterConnect_ThrowsOperationCanceledException

Result Details
Failed Saturday
Duration0:00:00.013
Failing build20201121.17
Ownernot available
Date started11/21/2020, 3:54:46 PM
Date completed11/21/2020, 3:54:46 PM
Failing sinceSaturday

Debug

Work items

Attachments

History

Comments
{ "HelixJobId": "f4526b9c-aaf7-4a96-af61-b4397413431b", "HelixWorkItemName": "System.Net.WebSockets.Client.Tests" }


Error message
System.IO.IOException : The operation is not allowed on non-connected sockets.


Stack trace
   at System.Net.Sockets.NetworkStream..ctor(Socket socket, FileAccess access, Boolean ownsSocket) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs:line 57
   at System.Net.Sockets.NetworkStream..ctor(Socket socket, Boolean ownsSocket) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs:line 32
   at System.Net.Test.Common.LoopbackServer.EstablishConnectionAsync() in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 108
   at System.Net.Test.Common.LoopbackServer.AcceptConnectionAsync(Func`2 funcAsync) in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 121
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 96
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 124
   at System.Net.Test.Common.LoopbackServer.<>c__DisplayClass13_0.<<CreateClientAndServerAsync>b__0>d.MoveNext() in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 87
--- End of stack trace from previous location ---
   at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options) in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 71
   at System.Net.WebSockets.Client.Tests.ConnectTest.ConnectAsync_CancellationRequestedAfterConnect_ThrowsOperationCanceledException() in /_/src/libraries/System.Net.WebSockets.Client/tests/ConnectTest.cs:line 264
--- End of stack trace from previous location ---
```
https://dev.azure.com/dnceng/public/_build/results?buildId=895674&view=ms.vss-test-web.build-test-results-tab&runId=28623020&paneView=debug&resultId=139893
net5.0-OSX-Debug-x64-CoreCLR_release-OSX.1013.Amd64.Open

@lewing is there a fix to backport?

**Comment by lewing** [comment]
Probably worth a look, the failure rate on release/5.0 vs main definitely suggests something got fixed.

**Comment by steveisok** [comment]
I'm not sure.  It seems to be hitting both runtimes.  Once in the last 7 days on mono and 4 for coreclr and they both started happening on the 21st.  https://github.com/dotnet/runtime/pull/45062 was the first instance.  Seems like a coincidence.

**Comment by wfurt** [comment]
seems similar to #38159 and my attempt to fix it in #38900. AFAIK OSX fails certain calls if peer already started close ;(
Perhaps we should disable it on release branch to avoid noise. 
And we track master failures weekly and it does not show there among top offenders.

**Comment by stephentoub** [comment]
Don't we just need to backport https://github.com/dotnet/runtime/pull/42054?  The stack trace here shows an IOException coming from the loopback server, and that PR catches/eats all such exceptions.

**Comment by karelz** [comment]
Fixed in 6.0 (master) in PR #42054 and in 5.0.x in PR #45201.
